VideoPlayer: implement drain for video codecs

FernetMenta opened the pull request:
Video codecs are supposed to react on DVD_CODEC_CTRL_DRAIN flag appropriately. If this flag is set by VideoPlayer via SetCodecControl(), codecs should submit remaining pics. Player will call codec again until codec signals empty by returning VC_BUFFER. codec should no block longer than 100ms.

This makes libmpeg2 obsolete. All codecs supporting drain can handle still frames.

popcornmix committed:
mmalcodec: Add SetCodecControl function

FernetMenta committed:
VideoPlayer: fix message queue signal level of zero when not empty

FernetMenta committed:
VideoPlayer: msg queue cosmetics

FernetMenta committed:
VideoPlayer: cosmetics

FernetMenta committed:
VideoPlayer: video - refactoring, cleanup, minor fixes

popcornmix committed:
mmalcodec: Switch to a condition variable when blocking waiting for a picture

popcornmix committed:
mmalcodec: Implement DVD_CODEC_CTRL_DRAIN to drain pictures

FernetMenta committed:
VideoPlayer: make ffmpeg and vaapi handle still frames

FernetMenta committed:
VideoPlayer: do not force software decoder for DVDs

FernetMenta committed:
VideoPlayer: drop libmpeg2

FernetMenta committed:
VideoPlayer: implement drain for vdpau

popcornmix commented:
Looking promising on Pi. Can you cherry-pick https://github.com/popcornmix/xbmc/commit/1bb13476620686aa2e18ecf2b61c1492a202c3d7 and https://github.com/popcornmix/xbmc/commit/5ce7a3911cb92054801910419da4363756178325 for drain support on mmal.

FernetMenta commented:
will do

FernetMenta commented:
@popcornmix could you send me PR please. one of the commits does not apply

FernetMenta committed:
VideoPlayer: cleanup codec factory

FernetMenta committed:
[droid] - adapt MediaCodec to factory cleanup

FernetMenta committed:
[IMX] - adapt video codec to factory cleanup

FernetMenta committed:
[iOS] - adapt vtb to factory cleanup

FernetMenta committed:
[aml] - adapt video codec to factory cleanup

FernetMenta committed:
[OpenMax] - adapt video codec to factory cleanup

FernetMenta commented:
I cleaned up the mess in factory codec. Codecs can decide on their own if they can handle the request or not. Currently ffmpeg (including hwaccel) and MMAL have implemented drain and can handle still frames.
@popcornmix @Memphiz @koying @wolfgar please check your codecs.

Memphiz commented:
@FernetMenta sorry for the dumb question - but what exactly do you want me to check? VTB is only used for h.264 - not sure how this would interfer with dvd playback and stills which is mpeg2 afaik?

Memphiz committed:
[VTB] - Check for supported codecs inside vtb decoder

FernetMenta commented:
@Memphiz I eliminated the ifdeffery in codec factory and moved the decision if the codec wants to take the job into the codecs. I want you to check if I missed anything: https://github.com/FernetMenta/xbmc/commit/b8982b2f96a360e8afa23285ad857b1f80b89aa6

I dropped CDVDVideoCodecVDA from factory. I think we should go for vtb on OSX too. Is that ok?

Memphiz commented:
thats ok i guess - the ffmpeg vda impl. we use seems still to have some issues on some GPUs though which was working with our VDA codec. VTB for osx would be a good thing i think @davilla might have more background to tell us if vtb for osx is what apple wants us to use atm?

fritsch commented:
Yes, it is the suggested way to go forward. It also makes big steps forward in ffmpeg itself,

popcornmix commented:
Just tried the Spongebob dvd and it the menus are fine on Pi with hardware decode enabled.

fritsch commented:
@popcornmix OSD controls work when the "final" video starts?

popcornmix commented:
There is no "feature" in the Spongebob DVD, but I was able to navigate into "audio options", "special features" and "episode selection". Which part did you mean by "final" video?

fritsch commented:
Lol - you are right, obviously. I left the screen too long alone and then the menu was gone (with the selection) and pressing return did nothing, yeah - cause there is no movie.

So all fine, sorry for the noise.

FernetMenta committed:
VideoPlayer: fix for snc of a/v streams - count packets received from demuxer

FernetMenta commented:
@fritsch branch drain was missing https://github.com/xbmc/xbmc/commit/4f2ae0cbebcadc85bf09e232ea082cc1be4559e6

@popcornmix the Alan Parsons DVD is now navigable with working time indicator

fritsch commented:
@FernetMenta we need a workaround for DXVA in FFMpeg Codec that opens it to not open for stills until a windows dev jumps on that ship, too.

FernetMenta commented:
@fritsch see my comment on the other commit. dxva can check for hint.stills and return false if true

fritsch commented:
It is not called like that: http://sprunge.us/WKTa should do I think

FernetMenta commented:
@fritsch https://github.com/FernetMenta/xbmc/commit/2c9d0ff4e4fd5f22297cd91822d1dcb80c922351

fritsch commented:
Thx. Btw. it is still a design issue as the decoders itself don't get this information. And decoder workarounds need to be implemented in FFmpegCodec. Something for another day.

FernetMenta commented:
@fritsch good point: https://github.com/FernetMenta/xbmc/commit/a8c3085dfe8055a639d6b2ee2d28e6c739780562

FernetMenta committed:
VideoPlayer: handle exceptions for caching

Voyager1 commented:
hi folks - thanks again for looking into this - and this on the last day of the year :-)
Anyway, I tried current code in this PR (plus compile fix for the protected hints member), and I'm not able to make the SPONGEBOB dvd example work correctly, even turned off HW decoding altogether. The stills no longer display at all. After pressing select 'in the dark' I get the moving menu items, but again once in the main menu I can't jump to the episode selection menu (which is still). The other thing I've noticed, is that the SPONGEBOB video is stretched in 16:9 despite the setting to keep things in 4:3.

I have uploaded another example (https://www.dropbox.com/s/p9x4v4ddps5trbs/WILLEM.zip?dl=0 ) that also exhibits the "green screen" phenomenon (mpeg2 that only decodes well in software). Apart from that, I also noticed:  
- much too slow fade-in on the warning screen
- can't navigate to the still menu (for episode selection)

fritsch commented:
Do you by chance use an intel gpu with outdated drivers? Or even worse Windows 10 on AMD?

Could you please also provide your debuglog? And the real question: Does it even work with v16 nightly?

Voyager1 commented:
@fritsch no my drivers are all up to date, but yes it's Win 10 x64 with AMD 15.12 drivers.
I have done testing with three versions:
- v16 nightly (d/l from the kodi site) - all works as expected
- v17 alpha Dec.1 (built from master) - SPONGEBOB misses the "swimming orange fish" animation intro before the language selection menu - all else works as expected
- v17 alpha Jan 1 (built from master + FernetMenta drain branch merged) - SPONGEBOB orange fish intro is back (!), but stills go to black. Have to blindly hit enter on the (black) language screen. The next two stills are black also, and then getting the dutch "brain" video and finally the spongebob animated menu. The menu is in 16:9 but it should be in 4:3. When navigating to the episode selection, the menu freezes (as it's a still again). With the other video, WILLEM, the "warning" video intro plays very slowly at 2 FPS. When in the menu, audio pitch is borked (goes up/down). When I play all (top menu item), the video plays as a "green screen".

Debug logs: https://www.dropbox.com/s/k9lnmgawtxuws3d/Kodi.zip?dl=0

FernetMenta commented:
The last commit fixes the 2 FPS "warning" intro. Then it is going to get interesting. I tracked it down to ffmpeg sw decoder. On OSX and Linux I get a picture for the still frame after having fed the decoder with 2 packets. On Windows 2 packets with exactly the same size go in and nothing comes out :(

FernetMenta commented:
I am trying to upgrade mingw with its rather old compiler. Seems to be no easy task ...

afedchin commented:
@FernetMenta look at this https://github.com/afedchin/xbmc/tree/msys2

Voyager1 commented:
@FernetMenta - I'm confident that most of the issues can be fixed somehow, but what about the mpeg2 that only decodes well in software (the green-screen video)? I'm pretty certain this will stay unresolved, even on Linux I guess. Shouldn't we also provide an option, next to System > Video > Acceleration settings > Hardware Acceleration currently YES/NO, add AUTO. I've seen this in VLC too, this "auto" uses HW accel only for HD video. This way we could satify the Pi's needs while maintaining maximum compatibility with legacy video.

FernetMenta commented:
@Voyager1 the green screen is not a decoder issue, this is rendering. When renderer tried to render something non existent the results look mostly green. I haven't seen any green screens with this branch on OSX and Linux so far.

Voyager1 commented:
@FernetMenta do me a favor and try the 'WILLEM' dvd on linux, and see for yourself if this decodes fine in hardware...

FernetMenta commented:
@afedchin that branch looks a bit outdated in regard to recent ffmpeg options. do you have something up-to-date?

FernetMenta commented:
@WILLEM plays fine on OSX. Trust me, that is a rendering issue.

afedchin commented:
@FernetMenta I'll update it today later.

Voyager1 commented:
> Trust me, that is a rendering issue

I like to believe you... :-) I only have a handful of DVDs that exhibit this issue... the majority of dvd mpeg2 is playing just fine with DXVA.

Memphiz commented:
osx doesn't use hw decoder for mpeg2 btw...

FernetMenta commented:
@Voyager1 if there's any issue left with ffmpeg, we'll fix it there.

Voyager1 commented:
I'm lost - OSX doesn't use hw decoder - so quite logically it works fine there... and if it's "only" a rendering issue, then why would there be an ffmpeg issue? It seems so much that hardware decoder is the sole differentiating factor. I'm going to upload one more example, a DVD rip without menu, just the main title.

Voyager1 commented:
here it is: reauthored, no menu, only 2 mins of main title - 115Mb download:
https://www.dropbox.com/s/0m47infk0qae132/MAGIC_EN.zip?dl=0

Voyager1 commented:
![Image](http://i64.tinypic.com/o0zevp.jpg)

FernetMenta commented:
@Voyager1 maybe some misunderstanding. I am investigating an issue with ffmpeg sw decoder that works fine all all platforms but Windows. dxva decoder is a different story and it is disabled for dvds anyway in this branch.

this picture is clearly a rendering issue cause by dxva. maybe decoder submits an invalid surface to renderer. but as mentioned this is a different story because this branch should not use dxva for DVDs.

Voyager1 commented:
@FernetMenta - did you see that the decoder is ff-mpeg2video-dxva ? This is a DVD.

Voyager1 commented:
It's true you have disabled DXVA for stills (although there we hit the ffmpeg issue in software). This is the main title, so DXVA (hw decoding) kicks in.

FernetMenta commented:
@Voyager1 we can simply set stills outside the menu block or define a new variable in hints.

Voyager1 commented:
finally.. mystery solved. DXVA class has a "CheckCompatibility" method, which starts by marking all non-h264 codecs as de facto "compatible". Further down there's a check which is actually never reached, to mark SD <= 576 as "incompatible". Simply moved up that check and that did the trick. You can cherry pick from https://github.com/Voyager1/xbmc/commit/6c27b74af63b8957b260c3e5467ffe2ce056bab9
I also reverted https://github.com/FernetMenta/xbmc/commit/d2efa8f82e397c17fbeea8a8356f97bca7602220 as that doesn't work anyway.

afedchin commented:
@Voyager1 do you mean cherry-pick this one  https://github.com/Voyager1/xbmc/commit/ff7124a61f18fadd7a4a4803ae17835d3b394fc2

Voyager1 commented:
> do you mean cherry-pick this one Voyager1@ff7124a

yep... I corrected the post when I saw the mistake :-)

afedchin commented:
@FernetMenta here you go https://github.com/afedchin/xbmc/commits/msys2_upd
1. I'm not sure what libdvd patched correctly
2. Corrently the gnutls-static is builded locally instead of using precompled version.

FernetMenta commented:
@afedchin thanks, great work!
I build this branch but I am still having issues with the spongebob sample. After the fish intro the menu is not displayed. Like on OSX or Linux 2 packets with exactly the same size go into ffmpeg decoder but on Window no pic is produced. I have no idea how to track this down further without being able to debug into this. Do you have an idea? Are you able to build ffmpeg with toolchain=msvc? That should make source code debugging possible, right?

FernetMenta committed:
VideoPlayer: add hint for DVDs and have dxva ignore DVDs

FernetMenta commented:
@Voyager1 I added the dvd flag to hints: https://github.com/FernetMenta/xbmc/commit/0c5cb6cfa9028190b982e4a7935f7fcbe3f8c400

then we have live tv still working with dxva

Voyager1 commented:
@FernetMenta - thanks. I was googling the forum and found this: http://forum.kodi.tv/showthread.php?tid=195089 - there are even some live TV channels with HW decoding problems... That said, current code in DXVA.cpp is incorrect (see https://github.com/Voyager1/xbmc/commit/ff7124a61f18fadd7a4a4803ae17835d3b394fc2) , thus should either be fixed as proposed (but that also disables mpeg2 for live TV) or the useless mpeg2 compatibility check should be taken out...

ace20022 commented:
@afedchin could you try to PR the libdvd* updates when you habe time? This would make angle selection work properly, I guess.

FernetMenta commented:
@Voyager1 if you agree I would take that code out from dxva. There would be more users suffering from disabled hw decoding for live tv than benefit from making those very rare material playable without changing settings.

FernetMenta commented:
@afedchin @ace20022 I suggest that libdvd should leave our tree and go to depends like we did with ffmpeg. We don't even need our own form for libdvd because we don't apply fixes to it.

Voyager1 commented:
@FernetMenta - I'm ok. Perhaps we can later think of a setting in HW acceleration something like "only for HD"... IMO there's no urgency though.

Voyager1 commented:
libdvd has specific xbmc additions and patches.

FernetMenta commented:
> libdvd has specific xbmc additions and patches.

we have not done anything to this code for a long time. do we know if those are still required with new versions of libdvd? did anybody try to submit those upstream?

afedchin commented:
@FernetMenta Yes, some months ago I've builded ffmepg with msvc following this instruction https://trac.ffmpeg.org/wiki/CompilationGuide/MSVC just need to prepare msys2 env with correct link.exe and with --enable-debug flag

@ace20022 I'm not sure what I've patched libdvd* correctly. I think @FernetMenta is right the libdvd should go to depends. Just need to decide are kodi specific patches still required?

FernetMenta commented:
@afedchin could you give me instructions what changes I need to apply to msys2?

ace20022 commented:
@FernetMenta @Voyager1 @afedchin yes it should leave your tree. I think the biggest difference was the search by time, but I saw that a search sector by time was added to 5.0.0

Voyager1 commented:
Yup, I agree in principle. But next to the time seeking, there are some patches that may need to go upstream first:
- dvdnav : get/set state (important for resume playback).
- dvdread : cached reading (not sure if essential), partial read and some fixes to make it work with our vfs
- dvdcss : cached reading (not sure if essential)

See: https://github.com/xbmc/xbmc/tree/master/lib/libdvd/patches

FernetMenta commented:
I suggest to set up repos like we have for ffmpeg. Much easier to track the patches. Upstream works with git too. What you think?

Voyager1 commented:
makes sense to me. They are on git://git.videolan.org/... and they have a mirror on github too (https://github.com/vlc-mirror).

afedchin commented:
@FernetMenta in enviroment from my branch (you should build mingw libs once) go to xbmc/project/BuildDependencies/msys64/ and add at the beginning of mingw32_shell.bat
```
call "%VS120COMNTOOLS%vsvars32.bat"
```
then run it and follow this steps
```
# source /local32/etc/profile.local
# pacman -S coreutils
# export PATH="/c/Program Files (x86)/Microsoft Visual Studio 12.0/VC/BIN/":$PATH
# cd /build/ffmpeg
# CFLAGS="-I/local32/include" CXXFLAGS="-I/local32/include" LDFLAGS="-L/local32/lib" \
./configure --toolchain=msvc --prefix=/xbmc/lib/win32/ffmpeg --disable-static --enable-shared \
--enable-debug --enable-asm --enable-yasm --disable-doc --enable-gpl --disable-gnutls \
--enable-w32threads --enable-shared --enable-memalign-hack --enable-postproc --enable-zlib \
--disable-static --disable-programs --disable-devices --disable-avdevice --disable-crystalhd \
--enable-muxer=spdif --enable-muxer=adts --enable-muxer=asf --enable-muxer=ipod \
--enable-muxer=ogg --enable-encoder=ac3 --enable-encoder=aac --enable-encoder=wmav2 \
--enable-encoder=libvorbis --enable-encoder=png --enable-encoder=mjpeg --enable-protocol=http \
--enable-runtime-cpudetect --cpu=i686 --enable-libdcadec --extra-cflags="-MDd" \
--extra-ldflags="/NODEFAULTLIB:libcmt"
# make -j4 && make install
```
copy dlls from xbmc/lib/win32/ffmpeg/bin to xbmc/system/players/VideoPlayer folder
Note: no **gnutls**, no **libdcadec**

FernetMenta commented:
@afedchin thanks and maximum respect for setting up this msys2. I am able to debug into ffmpeg on Windows :)

@Voyager1 @ace20022 seems you have the most experience with libdvd. Do you want to setup our repos?

Voyager1 commented:
@FernetMenta - Christmas break is over and I'm back to work :-( I may have some time over the next weekend or two to try this... Unless @ace20022 has more bandwidth?

FernetMenta committed:
VideoPlayer - bypass ffmpeg filter when draining

FernetMenta committed:
VideoPlayer: drop dead code from dxva decoder

FernetMenta commented:
@Voyager1 dead code is dropped 2572ad4
The reason for the different behaviour on Windows was a different setting for deinterlacing. fixed by 68eb9cc now the menu after the fish intro shows

are you ok with the pr now?

popcornmix commented:
@FernetMenta can you drop https://github.com/FernetMenta/xbmc/commit/9a103b37089a97008ea40da3cb40fdb4355dfa42 (I know I suggested it...)
I've had a closer look and the switch statement a bit later will return false for unsupported codecs, so the commit is not required.

Voyager1 commented:
@FernetMenta - great news. I'm going to do some quick testing tonight, but it looks like all my DVD issues have been taken care of. I'll confirm.

FernetMenta commented:
@popcornmix done

Voyager1 commented:
@FernetMenta - tests done - all is good

afedchin commented:
can someone test current master + this PR on windows. Just wait till main DVD menu loop plays again. I've got a heap corruption inside driver. Just need to check on other hardware.

FernetMenta commented:
@afedchin I rebased today in the afternoon so this pr should actually be master + this pr. I tested on Windows and did not notice any issues on my machine which is Alienware Alpha, Win 10.

afedchin commented:
@FernetMenta thanks. seems an issue of Intel drivers.

FernetMenta commented:
jenkins build this please

Voyager1 commented:
@afedchin @FernetMenta I haven't seen that kind of issues on AMD, all is working good here.

Voyager1 commented:
@FernetMenta - I found another stills-related dvd menu issue. The example this time is "TARZ" https://www.dropbox.com/s/j22rlb9ur8e4n47/TARZ.zip?dl=0 . You find it when choosing English > Bonus Material. The screen "flashes" and the menu options are invisible. If you press twice "down", then select, something plays, then go back to DVD menu. Now you see the menu options! 
I also tested with the non-drain branch and there it works ok.

Voyager1 commented:
@FernetMenta - regarding the aforementioned remaining stills issue, when I turn OFF deinterlacing, the menu works fine... this is probably useful info.

popcornmix commented:
TARZ behaves fine on Pi (presumably because deinterlace path is rather different).

FernetMenta commented:
@Voyager1 yes, that did help. here is the fix: e833fceb9d93ab2fc9dbc42f6db3a0b83545086e

FernetMenta commented:
jenkins build this please

Voyager1 commented:
looks good indeed, thanks. 
During re-testing I also noticed that the deinterlace mode "software blend" is now broken on stills (pink screens), but that is the least of my concerns, we may as well get rid of it since ffmpeg has better alternatives.

FernetMenta commented:
thanks for feedback. deinterlacing methods is indeed something that needs rework. from a user perspective this is something that only should work best without putting too much load on a system.

FernetMenta commented:
build passed on all platforms (only some binary addon failed on Windows). if no objections, I would merge tomorrow.

afedchin commented:
@FernetMenta @Voyager1 guys can you check on windows again. I need to be sure. Please select render method DXVA and play dvd main menu till it plays loop again.

ace20022 commented:
I habe a laptop Witz an Intel 4600 and can test it later. Ans specific DVD?

afedchin commented:
@ace20022 it doesn't matter

Voyager1 commented:
@afedchin I admit I didn't have dxva rendering turned on but will test tonight and let you know

FernetMenta commented:
I tested again and don't observe any issues regardless what render method I choose. @afedchin what exactly do you observe?

ace20022 commented:
All samples from @Voyager1 work. The only thing I noticed is that the osd freezes at 8 seconds for the first clip of the willem dvd.

ace20022 commented:
After this 8 seconds stop does not work or is delayed till the end of the clip I guess.

Voyager1 commented:
@ace20022 - fyi my uploads mostly focused on the menu navigation, main titles were removed or truncated (like on willem dvd) so it's normal to expect weird results there.

ace20022 commented:
Ofc, but it's not the main title, it's the first intro clip

afedchin commented:
@FernetMenta I got a heap corruption while releasing processor's view when DXVA::CSurfaceContext is releasing after renderer's buffers recreation. I cannot figure out why.
debug log:
```
Debug Print: CAnnouncementManager - Announcement: OnPlay from xbmc
Debug Print: GOT ANNOUNCEMENT, type: 1, from xbmc, message OnPlay
Debug Print: ffmpeg[1094]: [mjpeg] Changeing bps to 8
'Kodi.exe' (Win32): Loaded 'C:\Windows\SysWOW64\avrt.dll'. Symbols loaded.
Debug Print: CAESinkDirectSound::CheckPlayStatus: Resuming Playback
Debug Print: CWin32DirectSound::GetSpace - buffer underrun - W:7680, P:0, O:5760.
Debug Print: CRenderManager::Configure - change configuration. 720x480. display: 720x405. framerate: 0.00. format: YV12
Debug Print: 'Kodi.exe' (Win32): Loaded 'C:\Windows\SysWOW64\igd11dxva32.dll'. Cannot find or open the PDB file.
CRenderManager::Configure - timeout waiting for configure
Debug Print: CWinRenderer::SelectRenderMethod: Requested render method: 5
Debug Print: DXVA::CProcessorHD::InitProcessor - Initing Video Enumerator with params: 720x480.
Debug Print: DXVA::CProcessorHD::InitProcessor - Video processor has 3 rate conversion.
Debug Print: CVideoPlayerVideo::OutputPicture - failed to configure renderer
Debug Print: DXVA::CProcessorHD::InitProcessor - Video processor has 0x2ef feature caps.
...
Debug Print: DXVA::CProcessorHD::InitProcessor - Filter 1 has following params - max: 1000, min: 0, default: 100
Debug Print: VideoPlayer::Sync - Audio - pts: 320000.000000, cache: 319999.933243, totalcache: 579999.983311
Debug Print: DXVA::CProcessorHD::InitProcessor - Filter 2 has following params - max: 1800, min: -1800, default: 0
Debug Print: VideoPlayer::Sync - Video - pts: 0.000000, cache: 50000.000000, totalcache: 100000.000000
Debug Print: DXVA::CProcessorHD::InitProcessor - Filter 3 has following params - max: 1000, min: 0, default: 100
Debug Print: CVideoPlayerAudio - CDVDMsg::GENERAL_RESYNC(-100000.000000)
Debug Print: DXVA::CProcessorHD::CreateSurfaces - Creating 5 processor surfaces with format 103.
Debug Print: CDVDAudio::Resume - resume audio stream
Debug Print: ActiveAE - start sync of audio stream
Debug Print: DXVA::CProcessorHD::OpenProcessor - Creating processor with input format: (103).
Debug Print: CWinRenderer::SelectRenderMethod: Selected render method 3: DXVA
Debug Print: CRenderManager::Configure - 5
Debug Print: Activating window ID: 12005
Debug Print: ------ Window Deinit (MyVideoNav.xml) ------
Debug Print: ------ Window Init (VideoFullScreen.xml) ------
Debug Print: Loading skin file: VideoFullScreen.xml, load type: KEEP_IN_MEMORY
Debug Print: ActiveAE::SyncStream - average error of -56.784382, start adjusting
Debug Print: ActiveAE::SyncStream - average error -26.784382 below threshold of 30.000000
Debug Print: Load VideoFullScreen.xml: 58.06ms
Debug Print: Alloc resources: 58.75ms  (58.62 ms skin load)
Debug Print: CWinEventsWin32::WndProcWindow is active
Debug Print: CWinEventsWin32::WndProc: Focus switched to process C:\Program Files (x86)\Microsoft Visual Studio 12.0\Common7\IDE\devenv.exe
Debug Print: ------ Window Deinit (Pointer.xml) ------
Debug Print: created video buffer 0
Debug Print: created video buffer 1
Debug Print: created video buffer 2
Debug Print: created video buffer 3
Debug Print: created video buffer 4
Debug Print: CWinRenderer::UpdateVideoFilter - chosen scaling method 1 is not supported by renderer
Debug Print: CWinRenderer::CreateIntermediateRenderTarget: format 88
Debug Print: CVideoPlayerVideo - CDVDMsg::GENERAL_RESYNC(-100000.000000)
Debug Print: CJoystick::Initialize : No Joystick found
Debug Print: CWinEventsWin32::WndProcWindow is active
Debug Print: ------ Window Init (Pointer.xml) ------
Debug Print: CDVDClock::Discontinuity - CVideoPlayerAudio::OutputPacket - was:1044170.052521, should be:984426.619506, error:-59743.433016
Debug Print: CPullupCorrection: detected pattern of length 1: 40000.00, frameduration: 40000.000000
Debug Print: ------ Window Deinit (Pointer.xml) ------
Debug Print: CVideoPlayerVideo::CalcFrameRate framerate was:25.000000 calculated:25.000000
Debug Print: CRenderManager::Configure - change configuration. 720x480. display: 720x405. framerate: 25.00. format: YV12
Debug Print: CWinRenderer::SelectRenderMethod: Requested render method: 5
Debug Print: CRenderManager::Configure - timeout waiting for configure
Debug Print: CVideoPlayerVideo::OutputPicture - failed to configure renderer
Debug Print: ActiveAE::SyncStream - average error -691.221021 above threshold of 100.000000
Debug Print: DXVA::CProcessorHD::InitProcessor - Initing Video Enumerator with params: 720x480.
Debug Print: DXVA::CProcessorHD::InitProcessor - Video processor has 3 rate conversion.
....
Debug Print: DXVA::CProcessorHD::InitProcessor - Filter 3 has following params - max: 1000, min: 0, default: 100
Debug Print: DXVA::CProcessorHD::CreateSurfaces - Creating 5 processor surfaces with format 103.
Debug Print: DXVA::CProcessorHD::OpenProcessor - Creating processor with input format: (103).
Debug Print: CWinRenderer::SelectRenderMethod: Selected render method 3: DXVA
Debug Print: CRenderManager::Configure - 5
Debug Print: created video buffer 0
Debug Print: created video buffer 1
Debug Print: DXVA::CSurfaceContext::~CSurfaceContext - destructing surface context
HEAP[Kodi.exe]: Heap block at 1532D2A0 modified at 153ABBBB past requested size of 7e913
Kodi.exe has triggered a breakpoint.

The program '[4176] Kodi.exe' has exited with code 0 (0x0).
```

callstack:
```
 	ntdll.dll!_RtlpBreakPointHeap@4()	Unknown
 	ntdll.dll!_RtlpCheckBusyBlockTail@8()	Unknown
 	ntdll.dll!RtlpValidateHeapEntry()	Unknown
 	ntdll.dll!_RtlDebugFreeHeap@12()	Unknown
 	ntdll.dll!RtlpFreeHeap()	Unknown
 	ntdll.dll!_RtlFreeHeap@12()	Unknown
 	igd10iumd32.dll!681f4e88()	Unknown
 	[Frames below may be incorrect and/or missing, no symbols loaded for igd10iumd32.dll]	
 	igd10iumd32.dll!687ea6fc()	Unknown
 	igd10iumd32.dll!683975f7()	Unknown
 	igd10iumd32.dll!6839035a()	Unknown
 	d3d11.dll!NDXGI::CDeviceChild<IDXGIResource1,IDXGISwapChainInternal>::FinalRelease()	Unknown
 	d3d11.dll!CLayeredObject<NDXGI::CResource>::Release()	Unknown
 	d3d11_3SDKLayers.dll!ATL::CComPtr<struct ID3D11LayeredUseCounted>::operator=(struct ID3D11LayeredUseCounted *)	Unknown
 	d3d11_3SDKLayers.dll!NDebug::CDeviceChild<struct ID3D11BlendState1>::FinalRelease(void)	Unknown
 	d3d11_3SDKLayers.dll!CLayeredObject<class NDebug::CTexture2D>::~CLayeredObject<class NDebug::CTexture2D>(void)	Unknown
 	d3d11_3SDKLayers.dll!CLayeredObject<class NDebug::CTexture2D>::Release(void)	Unknown
 	d3d11.dll!CUseCountedObject<NOutermost::CDeviceChild>::UCDestroy()	Unknown
 	d3d11.dll!CUseCountedObject<NOutermost::CDeviceChild>::UCReleaseUse()	Unknown
 	d3d11.dll!CLockOwnerChild<class CDevice,0>::UCReleaseUse(void)	Unknown
 	d3d11.dll!CLayeredObjectWithCLS<class CVideoProcessorInputView>::~CLayeredObjectWithCLS<class CVideoProcessorInputView>(void)	Unknown
 	d3d11.dll!CLayeredObjectWithCLS<class CVideoProcessorInputView>::`vector deleting destructor'(unsigned int)	Unknown
 	d3d11.dll!CLayeredObjectWithCLS<class CVideoProcessorInputView>::Release(void)	Unknown
 	d3d11_3SDKLayers.dll!ATL::CComPtr<struct ID3D11LayeredUseCounted>::operator=(struct ID3D11LayeredUseCounted *)	Unknown
 	d3d11_3SDKLayers.dll!NDebug::CDeviceChild<struct ID3D11BlendState1>::FinalRelease(void)	Unknown
 	d3d11_3SDKLayers.dll!CLayeredObject<class NDebug::CVideoProcessorInputView>::~CLayeredObject<class NDebug::CVideoProcessorInputView>(void)	Unknown
 	d3d11_3SDKLayers.dll!CLayeredObject<class NDebug::CVideoProcessorInputView>::Release(void)	Unknown
 	d3d11.dll!CUseCountedObject<NOutermost::CDeviceChild>::Release()	Unknown
 	d3d11_3SDKLayers.dll!CLayeredObject<class NDebug::CBlendState>::CContainedObject::Release(void)	Unknown
>	Kodi.exe!DXVA::CSurfaceContext::Reset() Line 703	C++
 	Kodi.exe!DXVA::CSurfaceContext::~CSurfaceContext() Line 594	C++
 	Kodi.exe!DXVA::CSurfaceContext::`scalar deleting destructor'(unsigned int)	C++
 	Kodi.exe!IDVDResourceCounted<DXVA::CSurfaceContext>::Release() Line 40	C++
 	Kodi.exe!DXVA::CRenderPicture::~CRenderPicture() Line 744	C++
 	Kodi.exe!DXVA::CRenderPicture::`scalar deleting destructor'(unsigned int)	C++
 	Kodi.exe!IDVDResourceCounted<DXVA::CRenderPicture>::Release() Line 40	C++
 	Kodi.exe!DXVABuffer::~DXVABuffer() Line 137	C++
 	Kodi.exe!DXVABuffer::`scalar deleting destructor'(unsigned int)	C++
 	Kodi.exe!CWinRenderer::DeleteYV12Texture(int index) Line 919	C++
 	Kodi.exe!CWinRenderer::CreateYV12Texture(int index) Line 927	C++
 	Kodi.exe!CWinRenderer::ManageTextures() Line 113	C++
```

looks like the corruption in the driver, but I've tried with different drivers with same result. This also doesn't happen with regular movie. only in DVD menu

FernetMenta commented:
@afedchin I think this problem starts here:
CRenderManager::Configure - timeout waiting for configure

@popcornmix observed a similar issues. The render thread is stuck more than 4 seconds in GUIInfoManager::SetCurrentItem doing some database queries. Can you investigate where the render thread is stuck on your system?

afedchin commented:
@FernetMenta I don't think so, removing a lot of log messages fixes the timeout but result is same.

FernetMenta commented:
I tried on a Win7 machine with Intel HD 4000: no issues

FernetMenta commented:
@what dvds are you testing with? did you try willem and spongebob?

afedchin commented:
I've tried with Name of Rose (have only this). will try with willem. where I can get spongebob sample?

afedchin commented:
no issue with willem. Here is a sample of The Name of Rose https://www.dropbox.com/s/phxq5etobkryr13/NameOfRose.zip?dl=0 which causes an issue.

FernetMenta commented:
can't find the link. try with willem: https://www.dropbox.com/s/p9x4v4ddps5trbs/WILLEM.zip?dl=0

I noticed that yours has a different size and aspect ratio: change configuration. 720x480. display: 720x405

Voyager1 commented:
here it is again: https://www.dropbox.com/s/hquzbke1x6pvz4e/SPONGEBOB.zip?dl=0

FernetMenta commented:
No issues on my systems with Name of the Rose

Voyager1 commented:
@afedchin - you were experimenting with the mingw libs (libdvd / ffmpeg etc.) - did you rebuild them as per current master? just guessing as to what could cause this atm.

afedchin commented:
no issues with spongebob. strange, why NoR causes an issue. I see only one difference is a double configure at the end of menu loop:
CRenderManager::Configure - change configuration. 720x480. display: 720x405. framerate: 0.00. format: YV12
CRenderManager::Configure - change configuration. 720x480. display: 720x405. framerate: 25.00. format: YV12

FernetMenta commented:
this looks normal. does it work with pixel shaders?

afedchin commented:
yes it works. I've found what cause an issue.

FernetMenta commented:
>  I've found what cause an issue.

great! would you share your findings :)

afedchin commented:
convert_yuv420_nv12 https://github.com/xbmc/xbmc/blob/master/xbmc/utils/win32/memcpy_sse2.h#L111 causes a heap corruption because chroma width (360) is not 0x10 aligned for 720x480 dimention. In some cases it cause an exception while texture is releasing.

afedchin commented:
here is a fix #8757

afedchin commented:
@FernetMenta I've wrote on slack my findings in RenderManager which causes PreInit twice.

FernetMenta commented:
@afedchin I saw it, thanks. there are more things I need to fix for renderers.

is this pr good to go now?

afedchin commented:
yes, I think

Uukrull commented:
@FernetMenta. This fix for Android works: https://github.com/Uukrull/xbmc/commit/9cdb47e504bfe371f8e7924dd6223ebb9e94ff3f

Voyager1 commented:
@FernetMenta - the SW blend deinterlacing is indeed a case that is broken since the drain implementation. A fix (for your inspiration) is here (you probably want to do this better): https://github.com/Voyager1/xbmc/commit/2b1059fd61e89886f829eda9b18a0ba5407e7d36

FernetMenta commented:
@Voyager1 I would drop SW_BLEND which is IMO useless now. ffmpeg's yadif produces much better results.
